---
rule: Z002
title: Unused variable with value
enabled: true
---

# Z002: Unused variable with value

Detects variables that are assigned a value but never used. If you don't need the value, use `_` as the variable name instead of a named discard like `_foo`.

## Examples

### Bad

```zig
// expect: Z002
const _unused = getValue();
```

### Good

Use plain `_` to discard values:

```zig
const _ = getValue();
```

Or use `__` (double underscore) for intentional discards that might be used later:

```zig
const __ = getValue();
```

## Rationale

Named discards like `_unused` suggest the variable might be used later, but actually discarding a computed value is often a mistake. Using plain `_` makes the intent to discard explicit.

The double underscore `__` exception allows for intentional discards during development or debugging.
