---
rule: Z003
title: Parse error
enabled: true
---

# Z003: Parse error

Reports syntax errors that prevent the Zig parser from understanding the code. These must be fixed before other rules can be checked.

## Examples

### Bad

```zig
// expect: Z003
const x = ;
```

### Good

```zig
const x = 42;
```

## Rationale

Parse errors indicate invalid Zig syntax. The linter reports these so you can see all issues (syntax and style) in one place, rather than needing to run the compiler separately.
