---
rule: Z004
title: Prefer anonymous struct init with type annotation
enabled: true
---

# Z004: Prefer anonymous struct init with type annotation

Prefer `const x: T = .{}` over `const x = T{}`. The type annotation style is more consistent with Zig idioms and works better with type inference.

## Examples

### Bad

```zig
// expect: Z004
const point = Point{ .x = 0, .y = 0 };
```

```zig
// expect: Z004
const config = Config{};
```

### Good

```zig
const point: Point = .{ .x = 0, .y = 0 };
```

```zig
const config: Config = .{};
```

## Rationale

The type annotation style `const x: T = .{}` is preferred because:

1. It's consistent with how Zig handles type inference elsewhere
2. It makes the type visible at the declaration site
3. It works better when the type comes from a function return or generic
4. It's the style used in the Zig standard library
