---
rule: Z005
title: Type function names should be PascalCase
enabled: true
---

# Z005: Type function names should be PascalCase

Functions that return types should use PascalCase names, following Zig's convention that types are PascalCase.

## Examples

### Bad

```zig
// expect: Z005
fn arrayList(comptime T: type) type {
    return struct { items: []T };
}
```

```zig
// expect: Z005
fn hash_map(comptime K: type, comptime V: type) type {
    return struct {};
}
```

### Good

```zig
fn ArrayList(comptime T: type) type {
    return struct { items: []T };
}
```

```zig
fn HashMap(comptime K: type, comptime V: type) type {
    return struct {};
}
```

## Rationale

In Zig, types are always PascalCase. Since type functions return types, they should also be PascalCase. This makes it clear at the call site that the result is a type:

```zig
const MyList = ArrayList(u32);  // Clear that MyList is a type
```

This matches the Zig standard library conventions (`ArrayList`, `HashMap`, `BoundedArray`, etc.).
