---
rule: Z007
title: Duplicate import
enabled: true
---

# Z007: Duplicate import

Detects when the same module is imported multiple times in the same file.

## Examples

### Bad

```zig
// expect: Z007
const std = @import("std");
const std2 = @import("std");
```

### Good

Import each module once and access items through it:

```zig
const std = @import("std");
const Allocator = std.mem.Allocator;
```

## Rationale

Duplicate imports are redundant and can lead to confusion about which binding to use. They may also indicate copy-paste errors.
