---
rule: Z010
title: Prefer anonymous init when type is known
enabled: true
---

# Z010: Prefer anonymous init when type is known

When returning or passing a struct/enum value where the type is already known from context, use anonymous initialization (`.{}` or `.value`) instead of explicit type initialization (`Type{}` or `Type.value`).

## Examples

### Bad

```zig
// expect: Z010
fn getPoint() Point {
    return Point{ .x = 0, .y = 0 };
}
```

```zig
// expect: Z010
fn getStatus() Status {
    return Status.ok;
}
```

### Good

```zig
fn getPoint() Point {
    return .{ .x = 0, .y = 0 };
}
```

```zig
fn getStatus() Status {
    return .ok;
}
```

Also applies to function arguments:

```
process(.{ .x = 0, .y = 0 });
setStatus(.ok);
```

## Rationale

When the type is known from context (return type, parameter type), repeating it is redundant. Anonymous initialization is more concise and idiomatic Zig. This is especially valuable when types are renamedâ€”the anonymous form doesn't need updating.
