---
rule: Z011
title: Deprecated function call
enabled: true
---

# Z011: Deprecated function call

Detects calls to functions or methods marked as deprecated with `@"deprecated: ..."` doc comments.

## Examples

### Bad

```
// Calling deprecated function
std.mem.tokenize(u8, "a b", " ");
```

### Good

Use the non-deprecated replacement:

```
std.mem.tokenizeScalar(u8, "a b", ' ');
```

## Rationale

Deprecated functions are marked for removal in future versions. Using them creates technical debt and potential future breakage. This rule helps you find and update deprecated API usage before it becomes a problem.

The rule detects:
- Direct function calls to deprecated functions
- Method calls on types with deprecated methods
- Aliases to deprecated functions
- Deprecated type functions

Note: This rule requires semantic analysis and may not detect all deprecated calls when type information is unavailable.
