---
rule: Z012
title: Public function exposes private type
enabled: true
---

# Z012: Public function exposes private type

Detects public functions that return or accept private (non-pub) types in their signature. This creates an unusable API since callers cannot name the type.

## Examples

### Bad

```zig
const PrivateType = struct { value: u32 };

// expect: Z012
pub fn getPrivate() PrivateType {
    return .{ .value = 42 };
}
```

```zig
const InternalConfig = struct { debug: bool };

// expect: Z012
pub fn configure(config: InternalConfig) void {
    _ = config;
}
```

### Good

Make the type public if the function is public:

```zig
pub const PublicType = struct { value: u32 };

pub fn getPublic() PublicType {
    return .{ .value = 42 };
}
```

Or make the function private:

```zig
const PrivateType = struct { value: u32 };

fn getPrivate() PrivateType {
    return .{ .value = 42 };
}
```

## Rationale

A public function with a private type in its signature is effectively unusable from outside the module. Callers cannot:
- Declare variables of the return type
- Pass the correct parameter type
- Store the result in a typed container

This is almost always a mistakeâ€”either the function should be private, or the type should be public.
