---
rule: Z013
title: Unused import
enabled: true
---

# Z013: Unused import

Detects imports that are never used in the file.

## Examples

### Bad

```zig
// expect: Z013
const std = @import("std");
const unused = @import("unused.zig");

pub fn main() void {
    std.debug.print("hello\n", .{});
}
```

### Good

Remove unused imports:

```zig
const std = @import("std");

pub fn main() void {
    std.debug.print("hello\n", .{});
}
```

Public re-exports are not considered unused:

```zig
pub const utils = @import("utils.zig");
```

Imports used only in tests are allowed when the import is in a test block:

```zig
test "example" {
    const testing = @import("testing.zig");
    testing.check();
}
```

## Rationale

Unused imports add noise to the file and may indicate dead code or incomplete refactoring. Removing them keeps the dependency list accurate and the code clean.
