---
rule: Z014
title: Error set names should be PascalCase
enabled: true
---

# Z014: Error set names should be PascalCase

Error set type names should use PascalCase, following Zig's convention that all types are PascalCase.

## Examples

### Bad

```zig
// expect: Z014
const file_error = error{ NotFound, PermissionDenied };
```

### Good

```zig
const FileError = error{ NotFound, PermissionDenied };
```

Inline error sets in function signatures don't need names:

```zig
fn read() error{ NotFound, EndOfStream }![]u8 {
    // ...
}
```

## Rationale

Error sets are types in Zig, and all types should be PascalCase. This makes it consistent with struct, enum, and union naming, and clear that the identifier represents a type.
