---
rule: Z015
title: Public function exposes private error set
enabled: true
---

# Z015: Public function exposes private error set

Detects public functions that return a private (non-pub) error set. Callers cannot properly handle errors they cannot name.

## Examples

### Bad

```zig
const InternalError = error{ Failed, Timeout };

// expect: Z015
pub fn doWork() InternalError!void {
    return error.Failed;
}
```

### Good

Make the error set public:

```zig
pub const WorkError = error{ Failed, Timeout };

pub fn doWork() WorkError!void {
    return error.Failed;
}
```

Or use an inferred error set:

```zig
pub fn doWork() !void {
    return error.Failed;
}
```

Or make the function private:

```zig
const InternalError = error{ Failed, Timeout };

fn doWork() InternalError!void {
    return error.Failed;
}
```

## Rationale

When a public function returns a private error set, callers cannot:
- Match on specific errors by name
- Propagate errors with proper typing
- Document which errors to expect

This makes error handling difficult and error-prone. Either expose the error set or use an inferred error set.
