---
rule: Z016
title: Split compound assert
enabled: true
---

# Z016: Split compound assert

Detects `assert(a and b)` which should be split into separate assertions for better error diagnostics.

## Examples

### Bad

```zig
const std = @import("std");
// expect: Z016
fn check() void {
    std.debug.assert(a and b);
}
```

### Good

Split into separate assertions:

```zig
const std = @import("std");
fn check() void {
    std.debug.assert(a);
    std.debug.assert(b);
}
```

Note: `assert(a or b)` is **not** flagged because it has different semanticsâ€”it passes if either condition is true.

## Rationale

When `assert(a and b)` fails, you don't know which condition failed. Splitting into separate assertions gives you precise failure information:

```
// With assert(a and b):
// "assertion failed" - which one?

// With separate asserts:
// "assertion failed: x > 0" - clear!
```

This is especially valuable in tests and invariant checks where quick diagnosis matters.
