---
rule: Z017
title: Redundant try in return
enabled: true
---

# Z017: Redundant try in return

Detects `return try expr` where `try` is redundant. When returning an error union, `return expr` handles error propagation automatically.

## Examples

### Bad

```zig
// expect: Z017
fn readFile(path: []const u8) ![]u8 {
    const file = std.fs.cwd().openFile(path, .{});
    return try file.readToEndAlloc(allocator, max_size);
}
```

### Good

```zig
fn readFile(path: []const u8) ![]u8 {
    const file = std.fs.cwd().openFile(path, .{});
    return file.readToEndAlloc(allocator, max_size);
}
```

The `try` is still needed for non-return statements:

```zig
fn process() !void {
    const data = try getData();  // try needed here
    return transform(data);       // no try needed in return
}
```

## Rationale

In Zig, `return expr` already propagates errors when the function returns an error union. Adding `try` is redundant and adds visual noise. The only purpose of `try` is to convert an error union to its payload type for intermediate operations.
