---
rule: Z020
title: Inline @This() should be assigned to constant
enabled: true
---

# Z020: Inline @This() should be assigned to constant

Detects inline usage of `@This()` that should be assigned to a constant like `Self` first.

## Examples

### Bad

```zig
const MyStruct = struct {
    value: u32,

    // expect: Z020
    pub fn clone(self: @This()) @This() {
        return self;
    }
};
```

### Good

For named structs, use the name directly (see Z019):

```zig
pub const MyStruct = struct {
    value: u32,

    pub fn clone(self: MyStruct) MyStruct {
        return self;
    }
};
```

For anonymous structs, assign `@This()` to `Self` first (see Z022).

## Rationale

Inline `@This()` is verbose and harder to read than a named constant. Assigning it to `Self` (or the struct name) once at the top of the struct makes the code cleaner and follows Zig conventions.
