---
rule: Z023
title: Parameter order convention
enabled: true
---

# Z023: Parameter order convention

Function parameters should follow a consistent order: comptime parameters before runtime parameters, and special types (Allocator, Io) in a conventional position.

## Examples

### Bad

```zig
// expect: Z023
fn process(data: []u8, allocator: std.mem.Allocator) void {
    _ = data;
    _ = allocator;
}
```

```zig
// expect: Z023
fn parse(reader: std.io.Reader, allocator: std.mem.Allocator) void {
    _ = reader;
    _ = allocator;
}
```

### Good

```
// Allocator before data parameters
fn process(allocator: Allocator, data: []u8) void

// Comptime parameters first
fn GenericList(comptime T: type) type

// Self/receiver parameters allowed first
fn parse(self: *Self, allocator: Allocator) void
```

## Rationale

Consistent parameter ordering makes APIs predictable. The Zig standard library follows these conventions:

1. `self` receiver (for methods)
2. `comptime` parameters
3. `Allocator`
4. I/O types (Reader, Writer)
5. Other parameters

Following this order makes your APIs feel familiar to Zig developers.
