---
rule: Z027
title: Access declaration through type instead of instance
enabled: true
---

# Z027: Access declaration through type instead of instance

Detects accessing struct constants or static functions through an instance when they should be accessed through the type.

## Examples

### Bad

```
// Accessing struct constant through instance
var c: Counter = .{};
_ = c.max;         // Bad: should be Counter.max

// Calling static function through instance
const p: Point = .{};
_ = p.origin();    // Bad: should be Point.origin()
```

### Good

```
// Access through the type
_ = Counter.max;
_ = Point.origin();

// Instance methods (with self parameter) are fine
var c: Counter = .{};
c.increment();     // OK - has self parameter
```

Note: This rule requires semantic analysis to distinguish instance methods from static functions.

## Rationale

Accessing type-level declarations through instances is misleadingâ€”it suggests the value depends on the instance when it doesn't. Using the type name makes it clear the declaration is static and independent of any instance.

Note: This rule requires semantic analysis to distinguish instance methods from static functions.
