---
rule: Z030
title: deinit should set self.* = undefined
enabled: true
---

# Z030: deinit should set `self.* = undefined`

Detects `deinit` functions with pointer receivers that don't set `self.* = undefined`. This is a Zig convention to help catch use-after-free bugs during development.

## Examples

### Bad

Missing `self.* = undefined`:

```zig
// expect: Z030
const Foo = struct {
    data: []u8,
    allocator: std.mem.Allocator,

    fn deinit(self: *Foo) void {
        self.allocator.free(self.data);
    }
};
```

Early return without defer:

```zig
// expect: Z030
const Foo = struct {
    data: ?[]u8,
    allocator: std.mem.Allocator,

    fn deinit(self: *Foo) void {
        if (self.data == null) return;
        self.allocator.free(self.data.?);
        self.* = undefined;
    }
};
```

### Good

Set `self.* = undefined` at the end:

```zig
const Foo = struct {
    data: []u8,
    allocator: std.mem.Allocator,

    fn deinit(self: *Foo) void {
        self.allocator.free(self.data);
        self.* = undefined;
    }
};
```

Use `defer` to handle all code paths:

```zig
const Foo = struct {
    data: ?[]u8,
    allocator: std.mem.Allocator,

    fn deinit(self: *Foo) void {
        defer self.* = undefined;
        if (self.data) |d| {
            self.allocator.free(d);
        }
    }
};
```

Non-pointer receivers are not checked:

```zig
const Foo = struct {
    data: u32,

    fn deinit(self: Foo) void {
        _ = self;
    }
};
```

## Rationale

Setting `self.* = undefined` after deinitialization is a Zig idiom that helps catch use-after-free bugs:

1. **Debug safety**: In debug builds, accessing `undefined` memory triggers a panic, making bugs immediately visible
2. **Valgrind/ASan friendly**: Undefined memory is more likely to be detected by memory sanitizers
3. **Convention**: This pattern is used throughout the Zig standard library

When a `deinit` function has early returns, using `defer self.* = undefined;` ensures all code paths properly invalidate the struct.

## See Also

- [Zig Documentation on undefined](https://ziglang.org/documentation/master/#undefined)
- [Zig Style Guide](https://ziglang.org/documentation/master/#style-guide)
