---
rule: Z031
title: Avoid underscore prefix in identifiers
enabled: true
---

# Z031: Avoid underscore prefix in identifiers

Identifiers should not use underscore prefixes like `_privateField` or `_helperFunction`. This pattern, common in other languages to indicate privacy, is discouraged in Zig.

## Examples

### Bad

```zig
// expect: Z031
fn _privateHelper() void {}
```

```zig
// expect: Z031
const _internalValue: u32 = undefined;
```

### Good

```zig
fn privateHelper() void {}
```

```zig
const internal_value: u32 = undefined;
```

Single underscore for discarding values is allowed:

```zig
const _ = someFunctionWithSideEffects();
```

Double underscore prefix (for compiler/builtin internals) is allowed:

```zig
const __internal_value = 42;
```

## Rationale

From the Zig Style Guide:

> In Zig, there are no private fields, and this style guide recommends against pretending otherwise. Instead, fields should be named carefully based on their semantics and documentation should indicate how to use fields without violating data invariants.

The underscore prefix convention from other languages (like Python's `_private` or C's `__internal`) doesn't translate well to Zig because:

1. Zig has explicit `pub` for visibility - if something isn't `pub`, it's already private to the module
2. Using underscores for "private" fields creates a false sense of encapsulation
3. It makes names harder to read and type
4. The underscore has a specific meaning in Zig: discarding values (`_ = x`)

Instead of `_helper`, just use `helper` and don't export it with `pub`.
