---
rule: Z032
title: Acronyms should use standard casing
enabled: true
---

# Z032: Acronyms should use standard casing

Acronyms and initialisms in identifiers should follow standard Zig casing rules, not be written in all-caps. Even two-letter acronyms are subject to these conventions.

## Examples

### Bad

```zig
// expect: Z032
const XMLParser = struct {};
```

```zig
// expect: Z032
const HTTPSConnection = struct {};
```

```zig
// expect: Z032
fn parseXML() void {}
```

```zig
// expect: Z032
const IOError = struct {};
```

### Good

```zig
const XmlParser = struct {};
```

```zig
const HttpsConnection = struct {};
```

```zig
fn parseXml() void {}
```

```zig
const IoError = struct {};
```

Single uppercase letters followed by lowercase are fine:

```zig
fn readU32() u32 { return 0; }
```

## Rationale

From the Zig Style Guide:

> Acronyms, initialisms, proper nouns, or any other word that has capitalization rules in written English are subject to naming conventions just like any other word. Even acronyms that are only 2 letters long are subject to these conventions.

This means:
- `XML` becomes `Xml` in PascalCase or `xml` in camelCase
- `HTTP` becomes `Http` in PascalCase or `http` in camelCase
- `IO` becomes `Io` in PascalCase or `io` in camelCase

The standard library follows this convention (e.g., `std.json`, `std.http`, `std.io`).

## See Also

- [Z001](Z001.md) - Function names should be camelCase
- [Z005](Z005.md) - Type function names should be PascalCase
- [Z006](Z006.md) - Variable names should be snake_case
