---
rule: Z033
title: Avoid redundant words in identifiers
enabled: false
---

# Z033: Avoid redundant words in identifiers

Identifiers should avoid words that add no meaningful information, such as "Value", "Data", "Context", "Manager", "State", "utils", and "misc".

**This rule is disabled by default** as it can be controversial. Enable it in your `.ziglint.zon` if you want stricter naming enforcement.

## Examples

### Bad (when enabled)

```zig
const UserData = struct {
    name: []const u8,
};
```

```zig
const StateManager = struct {
    state: State,
};
```

```zig
const AppContext = struct {
    config: Config,
};
```

### Good

```zig
const User = struct {
    name: []const u8,
};
```

```zig
const StateStore = struct {
    state: State,
};
```

```zig
const App = struct {
    config: Config,
};
```

## Configuration

To enable this rule, add to your `.ziglint.zon`:

```
.{
    .rules = .{
        .Z033 = .{ .enabled = true },
    },
}
```

## Rationale

From the Zig Style Guide:

> Avoid these words in type names:
> - Value
> - Data
> - Context
> - Manager
> - State
> - utils, misc, or somebody's initials
>
> Everything is a value, all types are data, everything is context, all logic manages state. Nothing is communicated by using a word that applies to all types.

These words are so generic that they don't help readers understand what the type actually represents. Compare:

- `UserData` vs `User` - "Data" adds nothing
- `ConfigManager` vs `ConfigStore` or `ConfigRegistry` - "Manager" is vague
- `AppContext` vs `App` or `Runtime` - "Context" is overused

This rule is disabled by default because:
1. Some legitimate uses exist (e.g., wrapping external APIs)
2. Migration burden on existing codebases can be high
3. The "right" alternative name is often context-dependent
